client/state/data-getters/index.js

```diff
@@ -91,6 +91,55 @@ export const requestActivityLogs = ( siteId, filter, { freshness = 5 * 60 * 1000
 	);
 };
 
+const requestExternalContributorsId = siteId => `site-external-contributors-${ siteId }`;
+
+export const requestExternalContributors = siteId =>
+	requestHttpData(
+		requestExternalContributorsId( siteId ),
+		http( {
+			method: 'GET',
+			path: `/sites/${ siteId }/external-contributors`,
+			apiNamespace: 'wpcom/v2',
+		} ),
+		{
+			fromApi: () => data => [ [ requestExternalContributorsId( siteId ), data ] ],
+		}
+	);
+
+export const requestExternalContributorsAddition = ( siteId, userId ) => {
+	const requestId = requestExternalContributorsId( siteId );
+	const id = `${ requestId }-addition-${ userId }`;
+	return requestHttpData(
+		id,
+		http( {
+			method: 'POST',
+			path: `/sites/${ siteId }/external-contributors/add`,
+			apiNamespace: 'wpcom/v2',
+			body: { user_id: userId },
+		} ),
+		{
+			fromApi: () => data => [ [ requestId, data ] ],
+		}
+	);
+};
+
+export const requestExternalContributorsRemoval = ( siteId, userId ) => {
+	const requestId = requestExternalContributorsId( siteId );
+	const id = `${ requestId }-removal-${ userId }`;
+	return requestHttpData(
+		id,
+		http( {
+			method: 'POST',
+			path: `/sites/${ siteId }/external-contributors/remove`,
+			apiNamespace: 'wpcom/v2',
+			body: { user_id: userId },
+		} ),
+		{
+			fromApi: () => data => [ [ requestId, data ] ],
```

Comment:
what I'm assuming from this is that the call to `/remove` includes the full list of external contributors in its response and that in the same format as to `GET /external-contributors` because we're replacing the value of `externalContributors` with the response from this call

if that's the case it'd be best _at minimum_ to document that here so that it's not entirely opaque to the next person who comes around and is wondering what was _supposed_ to be happening here.

_at best_ we would create an explicit parser that will show the exact structure of the data and also trigger a failure if that data comes in the wrong shape. for instance, oftentimes a bug will unexpectedly remove the `_envelope` parameter from a request and then the data comes back in the wrong form and we pass that data deep into the application and then it crashes due to a `TypeError`.

Parsers in `httpData` are functions that take an input value (from the API response) and return a new value; in addition, if they throw an exception, then the request automatically transitions from a success to a failure and then the `error` value of the `http-data` will contain the exception itself.

```js
const parseExternalContributors = ( { contributors } ) =>
	contributors.map( ( { user_id: userId, display_name: name } ) => ( { userId, name } ) )

…

	fromApi: () => parseExternalContributors,
```

in this construction we know exactly which API response fields we are expecting and when that data comes back wrong we have a scale against which any random dev can compare and say, "oh, it's all wrapped in a `body` property but we're accessing as if the values we want are at the top level."

hope you find this stuff useful!